packages/ses/src/compartment-shim.js

```diff
@@ -83,13 +89,143 @@ const assertModuleHooks = compartment => {
+ * constructor, each of these has the same prototype.
+ * Because the prototype is shared among compartments, the shared constructor
+ * on the prototype must be powerless.
+ *
+ * Following this pattern, the shared compartment constructor is also powerless
+ * and simply throws an exception.
+ */
+export const SharedCompartmentConstructor = function Compartment() {
+  throw new TypeError(
+    `Cannot call or construct compartment with shared constructor`,
+  );
+};
+
+SharedCompartmentConstructor.toString = () => {
+  return 'function Compartment() { [shim code] }';
+};
+
+/** CompartmentPrototype
+ * Every compartment constructor shares this prototype.
+ * As with other primordials, the
  */
-export class Compartment {
-  constructor(endowments = {}, modules = {}, options = {}) {
+const CompartmentPrototype = {
+  constructor: SharedCompartmentConstructor,
+
+  get globalThis() {
+    return privateFields.get(this).globalObject;
+  },
+
+  /**
+   * @param {string} source is a JavaScript program grammar construction.
+   * @param {{
+   *   transforms: Array<Transform>,
+   *   sloppyGlobalsMode: bool,
+   * }} options.
+   */
+  evaluate(source, options = {}) {
+    // Perform this check first to avoid unecessary sanitizing.
+    if (typeof source !== 'string') {
+      throw new TypeError('first argument of evaluate() must be a string');
+    }
+
+    // Extract options, and shallow-clone transforms.
+    const { transforms = [], sloppyGlobalsMode = false } = options;
+    const localTransforms = [...transforms];
+
+    const {
+      globalTransforms,
+      globalObject,
+      globalLexicals,
+    } = privateFields.get(this);
+    const realmRec = getCurrentRealmRec();
+
+    return performEval(realmRec, source, globalObject, globalLexicals, {
+      globalTransforms,
+      localTransforms,
+      sloppyGlobalsMode,
+    });
+  },
+
+  module(specifier) {
+    if (typeof specifier !== 'string') {
+      throw new TypeError('first argument of module() must be a string');
+    }
+
+    assertModuleHooks(this);
+
+    const { exportsProxy } = getDeferredExports(
+      this,
+      privateFields.get(this),
+      moduleAliases,
+      specifier,
+    );
+
+    return exportsProxy;
+  },
+
+  async import(specifier) {
+    if (typeof specifier !== 'string') {
+      throw new TypeError('first argument of import() must be a string');
+    }
+
+    assertModuleHooks(this);
+
+    return load(privateFields, moduleAnalyses, this, specifier).then(() => {
+      const namespace = this.importNow(specifier);
+      return { namespace };
+    });
+  },
+
+  importNow(specifier) {
+    if (typeof specifier !== 'string') {
+      throw new TypeError('first argument of importNow() must be a string');
+    }
+
+    assertModuleHooks(this);
+
+    const moduleInstance = link(privateFields, moduleAliases, this, specifier);
+    moduleInstance.execute();
+    return moduleInstance.exportsProxy;
+  },
+
+  toString() {
```

Comment:
Btw, before #372 this one was also being silently removed due to absence from the whitelist.